### ISSUE TITLE
Is the i2c_core module required with a generic kernel version

### ISSUE THREAD

**Issue by creydr** [NVIDIA/gpu-operator#24]
### 1. Quick Debug Checklist
- [x] Are you running on an Ubuntu 18.04 node?
yes, 18.04.3 LTS

- [x] Are you running Kubernetes v1.13+?
yes, v1.14.3

- [x] Are you running Docker (>= 18.06) or CRIO (>= 1.13+)?
yes, Docker 18.09.7

- [x] Do you have `i2c_core` and `ipmi_msghandler` loaded on the nodes?
ipmi_msghandler, but not i2c_core (see my question)

```
$ lsmod | grep -i ipmi_msghandler
ipmi_msghandler        53248  4 ipmi_devintf,ipmi_si,nvidia,ipmi_ssif

$ lsmod | grep -i i2c_core
```

- [x] Did you apply the CRD (`kubectl describe specialresources --all-namespaces`)
yes
```
kubectl describe specialresources --all-namespaces
Name:         gpu
Namespace:    gpu-operator-resources
Labels:       <none>
Annotations:  kubectl.kubernetes.io/last-applied-configuration:
                {"apiVersion":"sro.openshift.io/v1alpha1","kind":"SpecialResource","metadata":{"annotations":{},"name":"gpu","namespace":"gpu-operator-res...
API Version:  sro.openshift.io/v1alpha1
Kind:         SpecialResource
Metadata:
  Creation Timestamp:  2020-01-09T07:24:17Z
  Generation:          1
  Resource Version:    18051258
  Self Link:           /apis/sro.openshift.io/v1alpha1/namespaces/gpu-operator-resources/specialresources/gpu
  UID:                 0bd37963-32b1-11ea-8893-005056002a8d
Spec:
  Scheduling Type:  None
Events:             <none>
```

#### TL;DR version
`i2c_core` seems to be only available for aws and azure kernels. Is it really needed to run GPU jobs on a generic kernel version?

#### Long version
My host runs on Ubuntu 18.04 with a 4.15.0-66-generic kernel version. When tried to load the i2c_core module, I did not find it later via `lsmod`. Then I saw that the i2c_core module is only available in kernel versions for aws or azure:
```
$ apt-file search 'i2c-core.ko'
linux-modules-4.15.0-1007-aws: /lib/modules/4.15.0-1007-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1009-aws: /lib/modules/4.15.0-1009-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1009-azure: /lib/modules/4.15.0-1009-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1010-aws: /lib/modules/4.15.0-1010-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1011-aws: /lib/modules/4.15.0-1011-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1012-azure: /lib/modules/4.15.0-1012-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1013-azure: /lib/modules/4.15.0-1013-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1014-azure: /lib/modules/4.15.0-1014-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1016-aws: /lib/modules/4.15.0-1016-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1017-aws: /lib/modules/4.15.0-1017-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1018-azure: /lib/modules/4.15.0-1018-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1019-aws: /lib/modules/4.15.0-1019-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1019-azure: /lib/modules/4.15.0-1019-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1020-aws: /lib/modules/4.15.0-1020-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1021-aws: /lib/modules/4.15.0-1021-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1021-azure: /lib/modules/4.15.0-1021-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1022-azure: /lib/modules/4.15.0-1022-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1023-aws: /lib/modules/4.15.0-1023-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1023-azure: /lib/modules/4.15.0-1023-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1025-aws: /lib/modules/4.15.0-1025-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1025-azure: /lib/modules/4.15.0-1025-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1027-aws: /lib/modules/4.15.0-1027-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1028-azure: /lib/modules/4.15.0-1028-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1029-aws: /lib/modules/4.15.0-1029-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1030-azure: /lib/modules/4.15.0-1030-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1031-aws: /lib/modules/4.15.0-1031-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1031-azure: /lib/modules/4.15.0-1031-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1032-aws: /lib/modules/4.15.0-1032-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1032-azure: /lib/modules/4.15.0-1032-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1033-aws: /lib/modules/4.15.0-1033-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1034-aws: /lib/modules/4.15.0-1034-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1035-aws: /lib/modules/4.15.0-1035-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1035-azure: /lib/modules/4.15.0-1035-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1036-azure: /lib/modules/4.15.0-1036-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1037-aws: /lib/modules/4.15.0-1037-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1037-azure: /lib/modules/4.15.0-1037-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1039-aws: /lib/modules/4.15.0-1039-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1040-aws: /lib/modules/4.15.0-1040-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1041-aws: /lib/modules/4.15.0-1041-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1043-aws: /lib/modules/4.15.0-1043-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1044-aws: /lib/modules/4.15.0-1044-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1045-aws: /lib/modules/4.15.0-1045-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1047-aws: /lib/modules/4.15.0-1047-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1048-aws: /lib/modules/4.15.0-1048-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.15.0-1050-aws: /lib/modules/4.15.0-1050-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.18.0-1006-aws: /lib/modules/4.18.0-1006-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.18.0-1006-azure: /lib/modules/4.18.0-1006-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.18.0-1007-aws: /lib/modules/4.18.0-1007-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.18.0-1007-azure: /lib/modules/4.18.0-1007-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.18.0-1008-aws: /lib/modules/4.18.0-1008-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.18.0-1008-azure: /lib/modules/4.18.0-1008-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.18.0-1011-aws: /lib/modules/4.18.0-1011-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.18.0-1011-azure: /lib/modules/4.18.0-1011-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.18.0-1012-aws: /lib/modules/4.18.0-1012-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.18.0-1013-aws: /lib/modules/4.18.0-1013-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.18.0-1013-azure: /lib/modules/4.18.0-1013-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.18.0-1014-azure: /lib/modules/4.18.0-1014-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.18.0-1016-aws: /lib/modules/4.18.0-1016-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.18.0-1017-aws: /lib/modules/4.18.0-1017-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.18.0-1018-aws: /lib/modules/4.18.0-1018-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.18.0-1018-azure: /lib/modules/4.18.0-1018-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.18.0-1019-azure: /lib/modules/4.18.0-1019-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.18.0-1020-aws: /lib/modules/4.18.0-1020-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.18.0-1020-azure: /lib/modules/4.18.0-1020-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.18.0-1023-azure: /lib/modules/4.18.0-1023-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.18.0-1024-azure: /lib/modules/4.18.0-1024-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-4.18.0-1025-azure: /lib/modules/4.18.0-1025-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-5.0.0-1011-aws: /lib/modules/5.0.0-1011-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-5.0.0-1012-aws: /lib/modules/5.0.0-1012-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-5.0.0-1012-azure: /lib/modules/5.0.0-1012-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-5.0.0-1014-aws: /lib/modules/5.0.0-1014-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-5.0.0-1014-azure: /lib/modules/5.0.0-1014-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-5.0.0-1016-aws: /lib/modules/5.0.0-1016-aws/kernel/drivers/i2c/i2c-core.ko
linux-modules-5.0.0-1016-azure: /lib/modules/5.0.0-1016-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-5.0.0-1018-azure: /lib/modules/5.0.0-1018-azure/kernel/drivers/i2c/i2c-core.ko
linux-modules-5.0.0-1020-azure: /lib/modules/5.0.0-1020-azure/kernel/drivers/i2c/i2c-core.ko
``` 

Anyhow the device-plugin runs and the GPUs on the node was detected:
```
kubectl -n gpu-operator-resources get po  
NAME                                       READY   STATUS      RESTARTS   AGE
nvidia-container-toolkit-daemonset-92jgq   1/1     Running     0          2d16h
nvidia-device-plugin-daemonset-lntd9       1/1     Running     0          2d16h
nvidia-device-plugin-validation            0/1     Completed   0          2d16h
nvidia-driver-daemonset-4ggb5              1/1     Running     0          2d16h
nvidia-driver-validation                   0/1     Completed   25         2d18h
```

```
kubectl -n gpu-operator-resources logs nvidia-device-plugin-daemonset-lntd9
2020/01/10 15:01:53 Loading NVML
2020/01/10 15:01:53 Fetching devices.
2020/01/10 15:01:53 Starting FS watcher.
2020/01/10 15:01:53 Starting OS watcher.
2020/01/10 15:01:53 Starting GRPC server
2020/01/10 15:01:53 Starting to serve on /var/lib/kubelet/device-plugins/nvidia.sock
2020/01/10 15:01:53 Registered device plugin with Kubelet
```

```
kubectl describe no gpu-node                      
Name:               gpu-node
Roles:              <none>
Labels:             beta.kubernetes.io/arch=amd64
                    beta.kubernetes.io/os=linux
                    ...
                    feature.node.kubernetes.io/kernel-config.NO_HZ=true
                    feature.node.kubernetes.io/kernel-config.NO_HZ_IDLE=true
                    feature.node.kubernetes.io/kernel-version.full=4.15.0-66-generic
                    feature.node.kubernetes.io/kernel-version.major=4
                    feature.node.kubernetes.io/kernel-version.minor=15
                    feature.node.kubernetes.io/kernel-version.revision=0
                    feature.node.kubernetes.io/memory-numa=true
                    feature.node.kubernetes.io/network-sriov.capable=true
                    feature.node.kubernetes.io/pci-102b.present=true
                    feature.node.kubernetes.io/pci-10de.present=true
                    feature.node.kubernetes.io/system-os_release.ID=ubuntu
                    feature.node.kubernetes.io/system-os_release.VERSION_ID=18.04
                    feature.node.kubernetes.io/system-os_release.VERSION_ID.major=18
                    feature.node.kubernetes.io/system-os_release.VERSION_ID.minor=04
                    kubernetes.io/arch=amd64
Taints:             <none>
Unschedulable:      false
Conditions:
  Type                 Status  LastHeartbeatTime                 LastTransitionTime                Reason                       Message
  ----                 ------  -----------------                 ------------------                ------                       -------
  NetworkUnavailable   False   Fri, 10 Jan 2020 15:59:46 +0100   Fri, 10 Jan 2020 15:59:46 +0100   CalicoIsUp                   Calico is running on this node
  MemoryPressure       False   Mon, 13 Jan 2020 08:35:01 +0100   Fri, 10 Jan 2020 15:59:27 +0100   KubeletHasSufficientMemory   kubelet has sufficient memory available
  DiskPressure         False   Mon, 13 Jan 2020 08:35:01 +0100   Sat, 11 Jan 2020 12:06:59 +0100   KubeletHasNoDiskPressure     kubelet has no disk pressure
  PIDPressure          False   Mon, 13 Jan 2020 08:35:01 +0100   Fri, 10 Jan 2020 15:59:27 +0100   KubeletHasSufficientPID      kubelet has sufficient PID available
  Ready                True    Mon, 13 Jan 2020 08:35:01 +0100   Fri, 10 Jan 2020 15:59:27 +0100   KubeletReady                 kubelet is posting ready status. AppArmor enabled
Addresses:
  InternalIP:  1.2.3.4
  Hostname:    gpu-node
Capacity:
 ...
 nvidia.com/gpu:                           1
 ...
Allocatable:
 ...
 nvidia.com/gpu:                           1
 ...
System Info:
 ...
 Kernel Version:             4.15.0-66-generic
 OS Image:                   Ubuntu 18.04.3 LTS
 Operating System:           linux
 Architecture:               amd64
 Container Runtime Version:  docker://18.9.7
 Kubelet Version:            v1.14.3
 Kube-Proxy Version:         v1.14.3
PodCIDR:                     10.233.69.0/24
Non-terminated Pods:
  Namespace                  Name                                          CPU Requests  CPU Limits  Memory Requests  Memory Limits  AGE
  ---------                  ----                                          ------------  ----------  ---------------  -------------  ---
  ...
  gpu-operator-resources     nvidia-container-toolkit-daemonset-92jgq      0 (0%)        0 (0%)      0 (0%)           0 (0%)         2d16h
  gpu-operator-resources     nvidia-device-plugin-daemonset-lntd9          0 (0%)        0 (0%)      0 (0%)           0 (0%)         2d16h
  gpu-operator-resources     nvidia-driver-daemonset-4ggb5                 0 (0%)        0 (0%)      0 (0%)           0 (0%)         2d16h
  kube-system                calico-node-vdlsg                             150m (0%)     300m (0%)   64M (0%)         500M (0%)      5d15h
  kube-system                kube-proxy-v82cg                              0 (0%)        0 (0%)      0 (0%)           0 (0%)         2d18h
  kube-system                nodelocaldns-wr7dx                            100m (0%)     0 (0%)      70Mi (0%)        170Mi (0%)     5d15h
  node-feature-discovery     gpu-feature-discovery-qtrjk                   0 (0%)        0 (0%)      0 (0%)           0 (0%)         2d16h
  node-feature-discovery     nfd-worker-w2ts5                              0 (0%)        0 (0%)      0 (0%)           0 (0%)         3d18h
  ...
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  Resource                                 Requests        Limits
  --------                                 --------        ------
  ...
  nvidia.com/gpu                           0               0
  ...
Events:                                    <none>
```

and a GPU test job (vectorAdd validation job) ran successfully:
```
[Vector addition of 50000 elements]
Copy input data from the host memory to the CUDA device
CUDA kernel launch with 196 blocks of 256 threads
Copy output data from the CUDA device to the host memory
Test PASSED
Done
```

So my question is, if the i2c_core module is really needed and if so, how it can be enabled on an non-aws or non-azure kernel.

### 2. Steps to reproduce the issue
Try to load the i2c_core module on a standard kernel e.g. one from the standard ubuntu image:

```
$ uname -r
-> e.g. 4.15.0-66-generic
$ modprobe -a i2c_core
$ lsmod | grep i2c_core
```

Thanks for your help.

**Comment by RenaudWasTaken** [comment]
Hello!

I believe with a generic kernel, the i2c_core kernel module is part of the default kernel modules and hence why it doesn't appear in the lsmod list.

So you are correct to assume that the i2c_core kernel module isn't required :)
